tests/e2e/ibc_callbacks_test.go

```diff
@@ -0,0 +1,200 @@
+package e2e_test
+
+import (
+	"encoding/json"
+	"fmt"
+	"testing"
+	"time"
+
+	wasmvmtypes "github.com/CosmWasm/wasmvm/v2/types"
+	ibcfee "github.com/cosmos/ibc-go/v8/modules/apps/29-fee/types"
+	ibctransfertypes "github.com/cosmos/ibc-go/v8/modules/apps/transfer/types"
+	channeltypes "github.com/cosmos/ibc-go/v8/modules/core/04-channel/types"
+	ibctesting "github.com/cosmos/ibc-go/v8/testing"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	sdkmath "cosmossdk.io/math"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	"github.com/CosmWasm/wasmd/app"
+	"github.com/CosmWasm/wasmd/tests/e2e"
+	wasmibctesting "github.com/CosmWasm/wasmd/x/wasm/ibctesting"
+	"github.com/CosmWasm/wasmd/x/wasm/types"
+)
+
+func TestIBCCallbacks(t *testing.T) {
+	// scenario:
+	// given two chains
+	//   with an ics-20 channel established
+	//   and an ibc-callbacks contract deployed on chain A and B each
+	// when the contract on A sends an IBCMsg::Transfer to the contract on B
+	// then the contract on B should receive a destination chain callback
+	//   and the contract on A should receive a source chain callback with the result (ack or timeout)
+	marshaler := app.MakeEncodingConfig(t).Codec
+	coord := wasmibctesting.NewCoordinator(t, 2)
+	chainA := coord.GetChain(wasmibctesting.GetChainID(1))
+	chainB := coord.GetChain(wasmibctesting.GetChainID(2))
+
+	actorChainA := sdk.AccAddress(chainA.SenderPrivKey.PubKey().Address())
+	oneToken := sdk.NewCoins(sdk.NewCoin(sdk.DefaultBondDenom, sdkmath.NewInt(1)))
+
+	path := wasmibctesting.NewPath(chainA, chainB)
+	path.EndpointA.ChannelConfig = &ibctesting.ChannelConfig{
+		PortID:  ibctransfertypes.PortID,
+		Version: string(marshaler.MustMarshalJSON(&ibcfee.Metadata{FeeVersion: ibcfee.Version, AppVersion: ibctransfertypes.Version})),
+		Order:   channeltypes.UNORDERED,
+	}
+	path.EndpointB.ChannelConfig = &ibctesting.ChannelConfig{
+		PortID:  ibctransfertypes.PortID,
+		Version: string(marshaler.MustMarshalJSON(&ibcfee.Metadata{FeeVersion: ibcfee.Version, AppVersion: ibctransfertypes.Version})),
+		Order:   channeltypes.UNORDERED,
+	}
+	// with an ics-20 transfer channel setup between both chains
+	coord.Setup(path)
+
+	// with an ibc-callbacks contract deployed on chain A
+	codeIDonA := chainA.StoreCodeFile("./testdata/ibc_callbacks.wasm").CodeID
+	contractAddrA := chainA.InstantiateContract(codeIDonA, []byte(`{}`))
+	require.NotEmpty(t, contractAddrA)
+
+	// and on chain B
+	codeIDonB := chainB.StoreCodeFile("./testdata/ibc_callbacks.wasm").CodeID
+	contractAddrB := chainB.InstantiateContract(codeIDonB, []byte(`{}`))
+	require.NotEmpty(t, contractAddrB)
+
+	// ExecuteMsg is the ibc-callbacks contract's execute msg
+	type ExecuteMsg struct {
+		ToAddress      string             `json:"to_address"`
+		ChannelID      string             `json:"channel_id"`
+		TimeoutSeconds wasmvmtypes.Uint64 `json:"timeout_seconds"`
```

Comment:
Since we only need this for the difference from now in seconds, we can also jut use uint64 and JSON numbers for this field. Even uint32 would be sufficient

Reply:
Done